### DISCUSSION TITLE
Запуск скрипта по результатам прослушанного

### DISCUSSION THREAD

**Discussion by Bedmit** [Chimildic/goofy#93] [Помощь с алгоритмом] [answered]
Возможна ли реализовать запуск скрипта, только если плейлист прослушан?
Вернее, я понимаю, что триггер запуска настраивается вне зависимости от содержания скрипта. Но возможна ли проверка в самом скрипте и дальнейшая его обработка по результатам этой проверки.

Как вариант, при ежечасном запуске скрипта, в нем проверяется количество прослушанных треков этого плейлиста с помощью общей истории прослушивания, и если это количество достигает определенного минимума, продолжается дальнейшая отработка скрипта.  В противном случае, отработка скрипта заканчивается и он остается неизменным.

Или как вариант, все входящие аргументы основной части скрипта равны нулю, а принимают какое-то определенное значение только по достижении проверяемого минимума.

**Comment by Chimildic** [comment, marked as answer]
Триггер возможен только по времени. Вызываемая функция ограничена [лимитами](https://chimildic.github.io/goofy/#/desc?id=Ограничения).
Для проверки условия используются блоки `if`. Чтобы завершить функцию раньше используется `return`. 

Касаемо примера, достаточно удалить историю и проверить сколько треков осталось. 
```js
let tracks = Source.getPlaylistTracks('', 'id');
Filter.removeTracks(tracks, RecentTracks.get());
if (tracks.length >= 10) {
    // если больше или равно 10, завершить
    return;
}
// иначе что-то делать
```
Еще напрашивается сокращение объема истории `RecentTracks.get(500)` или фильтр по относительной дате [rangeDateRel](https://chimildic.github.io/goofy/#/func?id=rangedaterel). На случай, если плейлист допускает некоторые повторы с течением времени.

Варианта с _входящими аргументами_ не понял.

**Reply by Bedmit** [reply]
Ух ты, как просто оказалось. Мне кажется для качественного развития Goofy уже нахватает ввода условий и циклов.

Про входящие аргументы боюсь не объясню, т.к. сам не до конца понимаю свою мысль. Да и при простоте первого варианта, остальные не важны.

**Reply by Chimildic** [reply]
> уже нахватает ввода условий и циклов

Какого ввода, куда?

**Reply by Bedmit** [reply]
> куда?

В скрипты конечно. Знай я, что например вышепоказанное условие так легко можно использовать в скрипте, уже давно этим бы воспользовался. 
Все скрипты я составляю прочитав и проанализировав возможности функций, представленных в  инструкции. Т.е. вот они кубики, а я уже решаю  в каком порядке их ставить друг на друга, что бы получилась нужная мне башня. Но вдруг оказывается, что кубики можно ставить не только друг на друга, а и скреплять их между собой сбоку. Это же может дать возможность постройки каких то кардинально других вариантов башни.

Может быть есть смысл как-то отобразить эти возможности в инструкции? Какой-то отдельной главой например. Или ссылкой на ветку в форуме. Чувствую проверка условия, не единственная доступная для большинства возможность программирования потенциально расширяющая использование Goofy. 

P. S. Единственное что печалит, это соотношение затраченных усилий по всей этой модернизации и ее объяснения таким как я , на количество тех, кто этим реально воспользуется.

**Reply by Maracumba** [reply]
> > уже нахватает ввода условий и циклов
> 
> Какого ввода, куда?
Наверно @Bedmit имеет ввиду объяснение на примерах для людей незнакомых с JS очень близко, а только после знакомства с Goofy, как можно комбинировать между собой функции: "условия" и "циклы" как кодить. "if" "else" и так далее.
Я такие вопросы задаю знающим, излагая свои идеи своим языком) Но иногда бывает неудобно уже просить) Спасибо ребятам за помощь.

**Reply by Chimildic** [reply]
Будто тайну раскрыл) Нет, не вдруг ведь. С самого начала известно, что это библиотека на языке программирования со всеми вытекающими. Только это отпугивает всех. Кажется, что становится лучше, когда у всего есть примеры и конечная задача лишь скопировать. Хоть кто-то пробует. 

Мне даже не лень будет что-то написать. Но в интернете тысячи уроков, курсов, примеров для `JavaScript`. Пользы от нового не будет никакой. Подразумевая те же условия и циклы. Лишь еще один однобокий взгляд. 
Задачки со Spotify только контекст использования, который ничего сверхъестественного не добавляет. 

Имею ввиду, что надо разделить язык и библиотеку при изучении. Делать это на своих местах. 
По большей части язык даст лишь ключевые слова и синтаксис. Все остальное будет библиотеками, встроенными или внешними. 

В шаблонах есть примеры с условиями в [радаре жанра](https://chimildic.github.io/goofy/#/template?id=Радар-жанра) и [исполнителе дня](https://chimildic.github.io/goofy/#/template?id=Исполнитель-дня). Циклы так вообще повсюду. 

И никакой клондайк в конце не ждет. Может станет короче и быстрее, но прям координально нового я не представляю.

**Reply by Bedmit** [reply]
Наверное все зависит от взгляда того, кто смотрит. 
Да, человек мало-мальски знакомый с основами программирования скорее всего понимает, что это всего-лишь частный случай целого. Что все принципы JavaScript работают и здесь. Для него какие-то моменты не отраженные в описании Goofy, но являющиеся основой работы языка, конечно не являются откровением.

Я же смотрю глазами полного невежды в этом вопросе. Для меня Goofy - это отдельный самодостаточный сервис, который я использую _исключительно_ на основе инструкции, конкретных примеров к каждой функции и понятных для меня кусков кода из скриптов других пользователей (в том числе и из шаблонов).  Я вижу, что в шаблонах встречаются неизвестные для меня элементы, но не использую их, т.к. просто не понимаю. Изучать ради них основы языка я тоже вряд-ли буду, по многим причинам. В лучшем случае спрошу что это значит. Так это я себя еще считаю самым пытливым из невежд. 

Получается вопрос в том - на кого больше ориентирован сервис и его описание. На более-менее подготовленных пользователей, или на любого человека способного просто прочитать инструкцию и скомбинировать уже готовые элементы в нужном ему порядке. 
Жалко только, что этих вот "любых человеков" не очень-то много и наблюдается. Что действительно ставит под сомнение необходимость добавления еще каких-то подробностей в и так уже очень подробное описание.

**Reply by Chimildic** [reply]
Странно выходит. Инструкции Goofy читаются, а инструкции JavaScript игнорируются. Хотя тема и мотивация про одно и тоже)

В результате, что хотят получить участники кружка?

**Reply by Bedmit** [reply]
Некоторые участники кружка, вступая в него ни как не ожидали, что возможно потребуется еще и дополнительное изучение чего-либо. Ладно уж с теми, кто влип в эту секту. А вот новобранцев такими откровениями точно не заманишь ))

**Comment by Bedmit** [comment]
Появилось время для изучения вопроса. С виду все просто и логика понятна, но почему-то не работает как надо.
Второй день пытаюсь понять, что я делаю или понимаю не так.

Взял для теста конкретный плейлист. Если он равен 40 трекам, обновление не происходит. В противном случае он обновляется.
```
function Test2() {
    let Test = Source.getPlaylistTracks('Bedtime', '6G6y4MozfbuDLEVt2azMa6');
  
  if (Test.lenght = 40) {
    console.log('Не будет обновлен.');
    return;
    }
    console.log('Будет обновлен.');
  }
```
Да, при 40 треках, плейлист не обновляется. Убрал из него пару треков, а результат тот же.
![2021-06-30](https://user-images.githubusercontent.com/76230194/123974483-4e471c80-d9c5-11eb-8de8-f6561c77182d.png)
![2021-06-30 (2)](https://user-images.githubusercontent.com/76230194/123974539-5737ee00-d9c5-11eb-800d-0600da42a1a2.png)
![2021-06-30 (1)](https://user-images.githubusercontent.com/76230194/123974567-5bfca200-d9c5-11eb-83ad-9cd8b7c77c00.png)

**Comment by Bedmit** [comment]
Аналогичная ситуация и например при выставлении условия `< 40`
При любом количестве треков, скрипт всегда переходит к "Будет обновлен"

**Reply by Chimildic** [reply]
Английское слово `длина` очень коварное для нас. 

Ошибка в том, что написано _leng**ht**_, а не корректное _leng**th**_. JavaScript динамический язык, поэтому нет ни подсказок, ни строгой проверки. В своем первом посте в этой теме я также написал с ошибкой.
- Первый вариант интерпретируется как ключ у объекта. Поскольку его нет у массива, происходит приведение к логическому типу _ложь_. Поэтому условие никогда не сработает.
- Второй вариант зарезервированное слово у массива, количество элементов в нем.

Другая ошибка - знак приравнивания `=`. При сравнении в условии используется двойное равенство `==`.

**Reply by Bedmit** [reply]
Да теперь все ок. Подозревал, что ошибка где-то на уровне опечатки, но ни как не мог понять где. Спасибо.
